internal/graph/parallelchecker.go

```diff
@@ -0,0 +1,120 @@
+package graph
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+
+	"github.com/authzed/spicedb/internal/dispatch"
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+type ParallelChecker struct {
+	toCheck       chan *v1.DispatchCheckRequest
+	c             dispatch.Check
+	g             *errgroup.Group
+	checkCtx      context.Context
+	subject       *core.ObjectAndRelation
+	maxConcurrent uint8
+	results       *tuple.ONRSet
+
+	dispatchCount       uint32
+	cachedDispatchCount uint32
+	depthRequired       uint32
+
+	mu sync.Mutex
+}
+
+func NewParallelChecker(ctx context.Context, c dispatch.Check, subject *core.ObjectAndRelation, maxConcurrent uint8) *ParallelChecker {
+	g, checkCtx := errgroup.WithContext(ctx)
+	toCheck := make(chan *v1.DispatchCheckRequest)
+	return &ParallelChecker{toCheck, c, g, checkCtx, subject, maxConcurrent, tuple.NewONRSet(), 0, 0, 0, sync.Mutex{}}
+}
+
+func (pc *ParallelChecker) AddResult(resource *core.ObjectAndRelation) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+	pc.addResultUnderLock(resource)
+}
+
+func (pc *ParallelChecker) DispatchCount() uint32 {
+	return pc.dispatchCount
+}
+
+func (pc *ParallelChecker) CachedDispatchCount() uint32 {
+	return pc.cachedDispatchCount
+}
+
+func (pc *ParallelChecker) DepthRequired() uint32 {
+	return pc.depthRequired
+}
+
+func (pc *ParallelChecker) addResultUnderLock(resource *core.ObjectAndRelation) {
```

Comment:
I think you can interpret this name to mean the opposite of what you actually mean; I will always have to come check the implementation. 

Maybe just drop these helpers?

Reply:
I was hoping to make it clear it needs to be under the lock... any ideas on how to name it?

Reply:
Since we need the lock for the stats update anyway, I figure we just keep it and do both changes